File: examples/JUCE - Marshall Blues Breaker/audio_analyzer.py
#!/usr/bin/env python3

import sys
import wave
import numpy as np
from pathlib import Path

def analyze_wav(filename):
    """Analyze WAV file and report audio quality metrics"""
    
    filepath = Path(filename)
    if not filepath.exists():
        print(f"Error: File not found: {filename}")
        return False
    
    print(f"\nReading file: {filename}")
    
    try:
        with wave.open(str(filepath), 'rb') as wav_file:
            n_channels = wav_file.getnchannels()
            sample_width = wav_file.getsampwidth()
            sample_rate = wav_file.getframerate()
            n_frames = wav_file.getnframes()
            
            print(f"Audio Properties:")
            print(f"  Channels:        {n_channels}")
            print(f"  Sample Rate:     {sample_rate} Hz")
            print(f"  Bit Depth:       {sample_width * 8} bits")
            print(f"  Frames:          {n_frames}")
            print(f"  Duration:        {n_frames / sample_rate:.2f} seconds")
            
            # Read all frames
            audio_bytes = wav_file.readframes(n_frames)
            
            # Convert to numpy array
            audio_array = np.frombuffer(audio_bytes, dtype=np.int16)
            
            # Reshape to channels
            if n_channels > 1:
                audio_array = audio_array.reshape((-1, n_channels))
                # Mix channels
                audio_mono = np.mean(audio_array, axis=1)
            else:
                audio_mono = audio_array
            
            # Normalize to [-1, 1]
            audio_float = audio_mono.astype(np.float32) / 32768.0
            
            # === ANALYSIS ===
            print("\n" + "="*80)
            print("GUITAR OUTPUT ANALYSIS")
            print("="*80)
            
            # Peak analysis
            peak = np.max(np.abs(audio_float))
            peak_db = 20 * np.log10(peak + 1e-10)
            
            # RMS
            rms = np.sqrt(np.mean(audio_float ** 2))
            rms_db = 20 * np.log10(rms + 1e-10)
            
            # DC offset
            mean_val = np.mean(audio_float)
            
            # Crest factor
            crest = peak / (rms + 1e-10)
            
            print("\nSIGNAL METRICS:")
            print(f"  Peak Level:      {peak:.6f} ({peak_db:+.2f} dB)")
            print(f"  RMS Level:       {rms:.6f} ({rms_db:+.2f} dB)")
            print(f"  DC Offset:       {mean_val:.8f}")
            print(f"  Crest Factor:    {crest:.2f}x")
            
            # Clipping analysis
            hard_clip = np.sum(np.abs(audio_float) >= 0.99)
            soft_clip = np.sum((np.abs(audio_float) >= 0.95) & (np.abs(audio_float) < 0.99))
            
            hard_clip_pct = (hard_clip / len(audio_float)) * 100
            soft_clip_pct = (soft_clip / len(audio_float)) * 100
            
            print("\nCLIPPING ANALYSIS:")
            print(f"  Hard Clipping:   {hard_clip:,} samples ({hard_clip_pct:.4f}%)")
            print(f"  Soft Clipping:   {soft_clip:,} samples ({soft_clip_pct:.4f}%)")
            
            if hard_clip_pct > 0.1:
                print("  STATUS: CLIPPING DETECTED - reduce level!")
            elif soft_clip_pct > 1.0:
                print("  STATUS: Soft clipping (acceptable for overdrive)")
            else:
                print("  STATUS: No clipping")
            
            # Distortion (THD approximation via waveform symmetry)
            half_size = len(audio_float) // 2
            sym_error = np.mean(np.abs(audio_float[:half_size] - audio_float[-half_size:][::-1]))
            thd_est = min(100, (sym_error / (rms + 1e-10)) * 100)
            
            print("\nDISTORTION:")
            print(f"  Estimated THD:   {thd_est:.2f}%")
            if thd_est < 5:
                print("  CHARACTER:       Clean")
            elif thd_est < 15:
                print("  CHARACTER:       Light distortion")
            elif thd_est < 50:
                print("  CHARACTER:       Heavy distortion (typical overdrive)")
            else:
                print("  CHARACTER:       Extreme distortion")
            
            # Noise floor
            sorted_mag = np.sort(np.abs(audio_float))
            noise_floor_idx = len(sorted_mag) // 10
            noise_floor = np.sqrt(np.mean(sorted_mag[:noise_floor_idx] ** 2))
            snr = 20 * np.log10((peak + 1e-10) / (noise_floor + 1e-10))
            
            print("\nNOISE FLOOR:")
            print(f"  SNR:             {snr:.2f} dB")
            if snr > 40:
                print("  QUALITY:         Excellent")
            elif snr > 20:
                print("  QUALITY:         Good")
            else:
                print("  QUALITY:         Poor")
            
            # Overall assessment
            print("\nFINAL ASSESSMENT:")
            score = 100
            issues = []
            
            if peak_db > 0:
                score -= 40
                issues.append("CRITICAL: Digital clipping (peak >= 0dB)")
            elif hard_clip_pct > 0.1:
                score -= 30
                issues.append("Hard clipping detected")
            elif soft_clip_pct > 2:
                score -= 10
                issues.append("Moderate soft clipping")
            
            if peak_db < -24:
                score -= 20
                issues.append("Signal too quiet (<-24dB)")
            
            if abs(mean_val) > 0.01:
                score -= 10
                issues.append("DC offset present")
            
            print(f"  Score:           {score}/100 ", end="")
            if score >= 90:
                print("(A+)")
            elif score >= 80:
                print("(A)")
            elif score >= 70:
                print("(B)")
            elif score >= 60:
                print("(C)")
            else:
                print("(F)")
            
            if not issues:
                print("  Status:          OK - no issues detected!")
            else:
                print("  Issues:")
                for issue in issues:
                    print(f"    - {issue}")
            
            print("\n" + "="*80)
            
            return True
            
    except wave.Error as e:
        print(f"Error reading WAV file: {e}")
        return False
    except Exception as e:
        print(f"Error: {e}")
        return False

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python audio_analyzer.py <wav_file>")
        sys.exit(1)
    
    success = analyze_wav(sys.argv[1])
    sys.exit(0 if success else 1)

File: examples/JUCE - Marshall Blues Breaker/audio_analyzer.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <iomanip>
#include <fstream>
#include <cstring>
#include <cstdint>

struct WAVHeader {
    char riffId[4];
    int32_t fileSize;
    char waveId[4];
    char fmtId[4];
    int32_t fmtSize;
    int16_t audioFormat;
    int16_t numChannels;
    int32_t sampleRate;
    int32_t byteRate;
    int16_t blockAlign;
    int16_t bitsPerSample;
    char dataId[4];
    int32_t dataSize;
};

class SimpleAudioAnalyzer {
private:
    std::vector<float> audioData;
    int sampleRate;
    
public:
    SimpleAudioAnalyzer() : sampleRate(44100) {}
    
    bool loadWAV(const std::string& filename) {
        std::ifstream file(filename, std::ios::binary);
        if (!file.is_open()) {
            std::cerr << "Error: Cannot open file: " << filename << std::endl;
            return false;
        }
        
        std::cout << "Reading file: " << filename << std::endl;
        
        WAVHeader header;
        file.read(reinterpret_cast<char*>(&header), sizeof(WAVHeader));
        
        if (!file.good()) {
            std::cerr << "Error: Cannot read WAV header" << std::endl;
            return false;
        }
        
        std::cout << "WAV Header Read:" << std::endl;
        std::cout << "  Channels: " << header.numChannels << std::endl;
        std::cout << "  Sample Rate: " << header.sampleRate << std::endl;
        std::cout << "  Bits Per Sample: " << header.bitsPerSample << std::endl;
        std::cout << "  Data Size: " << header.dataSize << " bytes" << std::endl;
        
        sampleRate = header.sampleRate;
        int numSamples = header.dataSize / (header.bitsPerSample / 8) / header.numChannels;
        audioData.resize(numSamples);
        
        std::cout << "Converting " << numSamples << " samples..." << std::endl;
        
        if (header.bitsPerSample == 16) {
            std::vector<int16_t> buffer(numSamples * header.numChannels);
            file.read(reinterpret_cast<char*>(buffer.data()), header.dataSize);
            
            for (int i = 0; i < numSamples; ++i) {
                float sample = 0.0f;
                for (int ch = 0; ch < header.numChannels; ++ch) {
                    sample += static_cast<float>(buffer[i * header.numChannels + ch]) / 32768.0f;
                }
                audioData[i] = sample / header.numChannels;
            }
        } else if (header.bitsPerSample == 32) {
            std::vector<int32_t> buffer(numSamples * header.numChannels);
            file.read(reinterpret_cast<char*>(buffer.data()), header.dataSize);
            
            for (int i = 0; i < numSamples; ++i) {
                float sample = 0.0f;
                for (int ch = 0; ch < header.numChannels; ++ch) {
                    sample += static_cast<float>(buffer[i * header.numChannels + ch]) / 2147483648.0f;
                }
                audioData[i] = sample / header.numChannels;
            }
        }
        
        file.close();
        return true;
    }
    
    void analyze() {
        std::cout << "\n" << std::string(80, '=') << std::endl;
        std::cout << "AUDIO ANALYSIS REPORT" << std::endl;
        std::cout << std::string(80, '=') << "\n" << std::endl;
        
        if (audioData.empty()) {
            std::cout << "ERROR: No audio data" << std::endl;
            return;
        }
        
        // Basic stats
        float peak = 0.0f;
        float rms = 0.0f;
        float mean = 0.0f;
        int hardClipCount = 0;
        int softClipCount = 0;
        
        for (const auto& sample : audioData) {
            float abs_sample = std::abs(sample);
            peak = std::max(peak, abs_sample);
            rms += sample * sample;
            mean += sample;
            
            if (abs_sample >= 0.99f) hardClipCount++;
            else if (abs_sample >= 0.95f) softClipCount++;
        }
        
        rms = std::sqrt(rms / audioData.size());
        mean /= audioData.size();
        
        float peakDb = 20.0f * std::log10(peak + 1e-10f);
        float rmsDb = 20.0f * std::log10(rms + 1e-10f);
        
        std::cout << "[BASIC METRICS]" << std::endl;
        std::cout << "Duration:            " << std::fixed << std::setprecision(2) 
                  << (audioData.size() / static_cast<float>(sampleRate)) << " seconds" << std::endl;
        std::cout << "Sample Rate:         " << sampleRate << " Hz" << std::endl;
        std::cout << "Total Samples:       " << audioData.size() << std::endl;
        std::cout << "Peak Level:          " << std::setprecision(4) << peak << " (" 
                  << std::setprecision(2) << peakDb << " dB)" << std::endl;
        std::cout << "RMS Level:           " << std::setprecision(4) << rms << " (" 
                  << std::setprecision(2) << rmsDb << " dB)" << std::endl;
        std::cout << "Mean (DC Offset):    " << std::setprecision(6) << mean << std::endl;
        std::cout << "Crest Factor:        " << std::setprecision(3) << (peak / (rms + 1e-10f)) << "x" << std::endl;
        
        std::cout << "\n[CLIPPING ANALYSIS]" << std::endl;
        float hardClipPercent = (hardClipCount / static_cast<float>(audioData.size())) * 100.0f;
        float softClipPercent = (softClipCount / static_cast<float>(audioData.size())) * 100.0f;
        std::cout << "Hard Clipping (>99%):   " << hardClipCount << " samples (" 
                  << std::setprecision(4) << hardClipPercent << "%)" << std::endl;
        std::cout << "Soft Clipping (95-99%): " << softClipCount << " samples (" 
                  << std::setprecision(4) << softClipPercent << "%)" << std::endl;
        
        if (hardClipPercent > 0.1f) {
            std::cout << "WARNING: Hard clipping detected - reduce input or output!" << std::endl;
        } else if (softClipPercent > 1.0f) {
            std::cout << "NOTE: Soft clipping present" << std::endl;
        } else {
            std::cout << "OK: No significant clipping" << std::endl;
        }
        
        // THD estimate
        std::cout << "\n[DISTORTION ANALYSIS]" << std::endl;
        float symError = 0.0f;
        int halfSize = audioData.size() / 2;
        for (int i = 0; i < halfSize; ++i) {
            symError += std::abs(audioData[i] - audioData[audioData.size() - 1 - i]);
        }
        symError /= halfSize;
        float thd = std::min(100.0f, (symError / (rms + 1e-10f)) * 100.0f);
        
        std::cout << "Estimated THD:       " << std::setprecision(2) << thd << "%" << std::endl;
        if (thd < 5.0f) {
            std::cout << "Quality: CLEAN (minimal distortion)" << std::endl;
        } else if (thd < 15.0f) {
            std::cout << "Quality: MODERATE DISTORTION" << std::endl;
        } else if (thd < 50.0f) {
            std::cout << "Quality: HIGH DISTORTION (typical for overdrive)" << std::endl;
        } else {
            std::cout << "Quality: EXTREME DISTORTION" << std::endl;
        }
        
        // Noise floor
        std::cout << "\n[NOISE ANALYSIS]" << std::endl;
        std::vector<float> sortedByMagnitude = audioData;
        std::sort(sortedByMagnitude.begin(), sortedByMagnitude.end(), 
                  [](float a, float b) { return std::abs(a) < std::abs(b); });
        
        float noiseFloor = 0.0f;
        int noiseCount = sortedByMagnitude.size() / 10;
        for (int i = 0; i < noiseCount; ++i) {
            noiseFloor += sortedByMagnitude[i] * sortedByMagnitude[i];
        }
        noiseFloor = std::sqrt(noiseFloor / noiseCount);
        float noiseDb = 20.0f * std::log10(noiseFloor + 1e-10f);
        float snr = 20.0f * std::log10((peak + 1e-10f) / (noiseFloor + 1e-10f));
        
        std::cout << "Noise Floor:         " << std::setprecision(6) << noiseFloor 
                  << " (" << std::setprecision(2) << noiseDb << " dB)" << std::endl;
        std::cout << "Signal-to-Noise:     " << std::setprecision(2) << snr << " dB" << std::endl;
        if (snr > 40.0f) {
            std::cout << "SNR Quality: EXCELLENT" << std::endl;
        } else if (snr > 20.0f) {
            std::cout << "SNR Quality: GOOD" << std::endl;
        } else {
            std::cout << "SNR Quality: POOR - significant noise" << std::endl;
        }
        
        // Overall assessment
        std::cout << "\n[OVERALL ASSESSMENT]" << std::endl;
        int score = 100;
        if (hardClipPercent > 0.1f) score -= 30;
        else if (softClipPercent > 1.0f) score -= 15;
        else if (peak < 0.1f) score -= 20;
        
        if (std::abs(mean) > 0.01f) score -= 10;
        
        std::cout << "Quality Score:       " << score << "/100" << std::endl;
        if (score >= 90) {
            std::cout << "Grade: A+ (EXCELLENT)" << std::endl;
        } else if (score >= 80) {
            std::cout << "Grade: A (VERY GOOD)" << std::endl;
        } else if (score >= 70) {
            std::cout << "Grade: B (GOOD)" << std::endl;
        } else if (score >= 60) {
            std::cout << "Grade: C (FAIR)" << std::endl;
        } else {
            std::cout << "Grade: F (POOR)" << std::endl;
        }
        
        std::cout << "\n" << std::string(80, '=') << std::endl;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: audio_analyzer <wav_file>" << std::endl;
        return 1;
    }
    
    SimpleAudioAnalyzer analyzer;
    if (!analyzer.loadWAV(argv[1])) {
        return 1;
    }
    
    analyzer.analyze();
    return 0;
}
